### DISCUSSION TITLE
Generate language version specific attributes in incremental generators `RegisterPostInitialisationOutput`

### DISCUSSION THREAD

**Discussion by latonz** [dotnet/roslyn#71299] [Q&A] [answered]
An incremental generator can use `RegisterPostInitialisationOutput` to output source that will be visible in the semantic model later in the generator pipeline. This is often used to output attribute definitions. There is no input available in `RegisterPostInitialisationOutput` (not even the language version of the compilation).
I wanted to migrate the attributes of our source generator from an external assembly to be generated by the source generator in `RegisterPostInitialisationOutput`. The attributes include newer language features such as generic attributes introduced in C# 11, while the generator has a minimum language version requirement of C# 9. Is there a way to output language version specific code while keeping it visible in the semantic model in later steps of the generator pipeline?

May be similar / related to https://github.com/dotnet/roslyn/issues/61094

### Workarounds
* Rely on the Roslyn version: would not work correctly with explicitly defined language versions
* Use a `<Generator>.targets` file, `DefineConstants` according to the used language version and use `#if` in the generated attributes. This could work but seems to be  complicated and error prone.
* Not use new language features at all

**Comment by jasonmalinowski** [comment]
There isn't a way to do this natively, and your workarounds are probably your best bets. I think we've had this request before, right @chsienki?

**Reply by chsienki** [reply, marked as answer]
I'm not sure if we've had this specific request, but definitely seen things similar to it, mainly about allowing generated code to target the highest langver by default. That works when the compiler is up to date but people are using an explicit lower language version, but doesn't work if people are targeting different compiler versions.

In general though, I suppose we (compiler team) might want to consider making the lang version available as a constant for this purpose as it seems useful.

I see two options depending on the scenario:

1) Always a new compiler, sometimes a lower langver:
 - Do the targets idea above and map the lang version into a constant you can reference. Annoying that you have to do it yourself, but i think it should be pretty solid. Changing the language version in the project will change the parse options which will cause the PostInit sources to be reparsed, so it should just work as you expect

2) Various compiler versions:
 - It is possible to multi-target generators based on Roslyn version, so you could theoretically have a generator that emits C# 9 for compilers lower than a certain version, and one that uses C#11 for higher versions. This is really designed for using different API surfaces, rather than language versions and is a pretty big hammer, but might work depending on your situation.

**Reply by latonz** [reply]
Thanks for your prompt answers & explanations, the target file approach works. However, if the language version would be available as a constant, things would be much easier.
